pkg/utils/ldap/reconnect.go

```diff
@@ -0,0 +1,317 @@
+// Copyright 2022 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package ldap
+
+// LDAP automatic reconnection mechanism, inspired by:
+// https://gist.github.com/emsearcy/cba3295d1a06d4c432ab4f6173b65e4f#file-ldap_snippet-go
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-ldap/ldap/v3"
+	"github.com/rs/zerolog"
+)
+
+var (
+	errMaxRetries = errors.New("max retries")
+)
+
+type ldapConnection struct {
+	Conn  *ldap.Conn
+	Error error
+}
+
+// ConnWithReconnect maintains an LDAP Connection that automatically reconnects after network errors
+type ConnWithReconnect struct {
+	conn    chan ldapConnection
+	reset   chan *ldap.Conn
+	retries int
+	logger  *zerolog.Logger
+}
+
+// Config holds the basic configuration of the LDAP Connection
+type Config struct {
+	URI          string
+	BindDN       string
+	BindPassword string
+	TLSConfig    *tls.Config
+}
+
+// NewLDAPWithReconnect Returns a new ConnWithReconnect initialized from config
+func NewLDAPWithReconnect(config Config) *ConnWithReconnect {
+	conn := ConnWithReconnect{
+		conn:    make(chan ldapConnection),
+		reset:   make(chan *ldap.Conn),
+		retries: 1,
+	}
+	logger := zerolog.Nop()
+	conn.logger = &logger
+	go conn.ldapAutoConnect(config)
+	return &conn
+}
+
+// SetLogger sets the logger for the current instance
+func (c *ConnWithReconnect) SetLogger(logger *zerolog.Logger) {
+	c.logger = logger
+}
+
+func (c *ConnWithReconnect) retry(retries int, fn func(c ldap.Client) error) error {
+	conn, err := c.getConnection()
+
+	if err != nil {
+		return err
+	}
+
+	for try := 0; try <= retries; try++ {
+		if try > 0 {
+			c.logger.Debug().Msgf("retrying attempt %d", try)
+			conn, err = c.reconnect(conn)
+			if err != nil {
+				// reconnection failed stop this attempt
+				return err
```

Comment:
That will make it stop retrying if the 2nd reconnect resulted in an error, not? That isn't what we want, is it?

Reply:
Yes. For the moment that is ok. I am most interested in covering the case where we need to recover after the server closed the connection (e.g. because it was restarted or some idle timeout kicked in).

We can always add more sophisticated reconnection login later. (waiting with exponential backoff and so forth)

Reply:
Ah, I see. That also explains the hardcoded `retries: 1` in the comment below. 

Ok then, scratch both of them. We'll do it when we improve the reconnection logic.